### PR TITLE
Add methods for batch withdrawals requests and claiming

### PR DESCRIPTION
**Pull request by Psirex**

### CODE REVIEW

File: contracts/0.8.9/WithdrawalQueue.sol [hunk 1]
```diff
@@ -277,21 +279,79 @@ contract WithdrawalQueue is AccessControlEnumerable, WithdrawalQueueBase {
         return _requestWithdrawalWstETH(_amountOfWstETH, _recipient);
     }
 
-    /// @notice Request withdrawal of the provided token amount in a batch
-    /// NB: Always reverts
-    function requestWithdrawalBatch(uint256[] calldata, address[] calldata)
+    struct WithdrawalRequestInput {
```
Comment (TheDZhon) [inline contracts/0.8.9/WithdrawalQueue.sol:282]: nice idea :+1:

File: contracts/0.8.9/WithdrawalQueue.sol [hunk 2]
```diff
@@ -277,21 +279,79 @@ contract WithdrawalQueue is AccessControlEnumerable, WithdrawalQueueBase {
...
+    }
+
+    /// @notice Request the sequence of withdrawals according to passed `withdrawalRequestInputs` data using EIP-2612 Permit
+    /// @param _withdrawalRequestInputs an array of `WithdrawalRequestInput` data. The standalone withdrawal request will
+    ///  be created for each item in the passed list. If `WithdrawalRequestInput.recipient` is set to `address(0)`,
+    ///  `msg.sender` will be used as recipient.
+    /// @return requestIds an array of the created withdrawal requests
+    function requestWithdrawalsWithPermit(
+        WithdrawalRequestInput[] memory _withdrawalRequestInputs,
+        Permit[] memory _permits
```
Comment (TheDZhon) [inline contracts/0.8.9/WithdrawalQueue.sol:319]: not sure that we need an array of permits, tbh
maybe just one for the whole sum?

File: contracts/0.8.9/WithdrawalQueue.sol [hunk 3]
```diff
@@ -277,21 +279,79 @@ contract WithdrawalQueue is AccessControlEnumerable, WithdrawalQueueBase {
...
+            claimWithdrawal(_requestIds[i], _hints[i]);
+        }
+    }
+
+    /// @notice Returns the list of hints for the given request ids
+    /// @param _requestIds ids of the requests to get hints for
+    function findClaimHints(uint256[] calldata _requestIds, uint256 _firstIndex, uint256 _lastIndex) external view returns (uint256[] memory hintIds) {
+        hintIds = new uint256[](_requestIds.length);
+        for (uint256 i = 0; i < _requestIds.length; ++i) {
+            hintIds[i] = findClaimHint(_requestIds[i], _firstIndex, _lastIndex);
```
Comment (TheDZhon) [inline contracts/0.8.9/WithdrawalQueue.sol:353]: if order is enforced, we can start the search from the previously found hint

File: contracts/0.8.9/WithdrawalQueue.sol [hunk 4]
```diff
@@ -356,4 +443,26 @@ contract WithdrawalQueue is AccessControlEnumerable, WithdrawalQueueBase {
 
         return _recipient;
     }
+
+    modifier onlyValidWithdrawalRequestInput(WithdrawalRequestInput memory _withdrawalRequestInput) {
+        address token = _withdrawalRequestInput.token;
+        if (token != address(STETH) && token != address(WSTETH)) {
+            address[] memory withdrawableTokens = new address[](2);
```
Comment (TheDZhon) [inline contracts/0.8.9/WithdrawalQueue.sol:450]: maybe too sophisticated error handling

**Review by TheDZhon — COMMENTED** [review]
Thank you! :+1: 
Left some considerations :eyes:

File: contracts/0.8.9/WithdrawalQueue.sol (hunk 4 above)
Comment (Psirex) [reply contracts/0.8.9/WithdrawalQueue.sol]: Fixed. This check and the modifier were removed due to the methods interface change.

File: contracts/0.8.9/WithdrawalQueue.sol (hunk 2 above)
Comment (Psirex) [reply contracts/0.8.9/WithdrawalQueue.sol]: Fixed. Made batch methods more strict:
- standalone methods for stETH and wstETH
- one permit expected for the whole withdrawal amount

File: contracts/0.8.9/WithdrawalQueue.sol [hunk 5]
```diff
@@ -274,21 +276,113 @@ contract WithdrawalQueue is AccessControlEnumerable, WithdrawalQueueBase {
...
+        uint256 amount;
+        address recipient;
+    }
+
+    /// @notice Request the sequence of stETH withdrawals according to passed `withdrawalRequestInputs` data
+    /// @param _withdrawalRequestInputs an array of `WithdrawalRequestInput` data. The standalone withdrawal request will
+    ///  be created for each item in the passed list. If `WithdrawalRequestInput.recipient` is set to `address(0)`,
+    ///  `msg.sender` will be used as recipient.
+    /// @return requestIds an array of the created withdrawal requests
+    function requestWithdrawalBatch(WithdrawalRequestInput[] memory _withdrawalRequestInputs)
```
Comment (folkyatina) [inline contracts/0.8.9/WithdrawalQueue.sol:289]: Why it's not a calldata here?

File: contracts/0.8.9/WithdrawalQueue.sol (hunk 1 above)
Comment (folkyatina) [reply contracts/0.8.9/WithdrawalQueue.sol:279]: Is there any drawbacks in reusing these structs in all the methods?

File: contracts/0.8.9/WithdrawalQueue.sol [hunk 6]
```diff
@@ -274,21 +276,113 @@ contract WithdrawalQueue is AccessControlEnumerable, WithdrawalQueueBase {
...
+    ///  be created for each item in the passed list. If `WithdrawalRequestInput.recipient` is set to `address(0)`,
+    ///  `msg.sender` will be used as recipient.
+    /// @return requestIds an array of the created withdrawal requests
+    function requestWithdrawalBatchWithPermit(WithdrawalRequestInput[] memory _withdrawalRequestInputs, Permit memory _permit)
+        external
+        whenResumed
+        returns (uint256[] memory requestIds)
+    {
+        STETH.permit(msg.sender, address(this), _permit.value, _permit.deadline, _permit.v, _permit.r, _permit.s);
+        return requestWithdrawalBatchWstETH(_withdrawalRequestInputs);
```
Comment (folkyatina) [inline contracts/0.8.9/WithdrawalQueue.sol:343]: `requestWithdrawalBatch`?

File: contracts/0.8.9/WithdrawalQueue.sol [hunk 7]
```diff
@@ -274,21 +276,113 @@ contract WithdrawalQueue is AccessControlEnumerable, WithdrawalQueueBase {
...
     }
 
     /// @notice Claim withdrawals batch once finalized (claimable)
-    /// NB: Always reverts
-    function claimWithdrawalBatch(uint256[] calldata /*_requests*/ ) external pure {
-        revert Unimplemented();
+    /// @param _requestIds ids of the finalized requests to claim
+    /// @param _hints rate indices found offchain that should be used for claiming
+    function claimWithdrawals(uint256[] calldata _requestIds, uint256[] calldata _hints) external {
```
Comment (folkyatina) [inline contracts/0.8.9/WithdrawalQueue.sol:363]: Imho, we should go with batches or with plurals for request and claim methods, but not mixed.

File: contracts/0.8.9/WithdrawalQueue.sol (hunk 1 above)
Comment (folkyatina) [reply contracts/0.8.9/WithdrawalQueue.sol:279]: Or we can even get rid of the singular methods at all. I like this idea more and more.

**Review by folkyatina — COMMENTED** [review]
I think we should get rid of singular request/claim in favor of plural and get rid of the `Batch` in the names as well. It will give a smaller surface and more consistency.

File: contracts/0.8.9/WithdrawalQueue.sol (hunk 5 above)
Comment (Psirex) [reply contracts/0.8.9/WithdrawalQueue.sol]: Fixed. `memory` stayed here from the previous version, where it was more gas efficient due to passing structs into the internal methods.

File: contracts/0.8.9/WithdrawalQueue.sol (hunk 6 above)
Comment (Psirex) [reply contracts/0.8.9/WithdrawalQueue.sol]: Fixed.

File: contracts/0.8.9/WithdrawalQueue.sol (hunk 7 above)
Comment (Psirex) [reply contracts/0.8.9/WithdrawalQueue.sol]: Fixed. Removed non-batch versions of the methods for withdrawals requesting and claiming.

File: contracts/0.8.9/WithdrawalQueue.sol [hunk 8]
```diff
@@ -203,92 +205,118 @@ contract WithdrawalQueue is AccessControlEnumerable, WithdrawalQueueBase {
...
-        external
-        view
-        whenResumed
-        returns (uint256[] memory)
-    {
-        revert Unimplemented();
+    struct ClaimWithdrawalInput {
+        /// @notice id of the finalized requests to claim
+        uint256 requestId;
+        /// @notice rate index found offchain that should be used for claiming
```
Comment (TheDZhon) [inline contracts/0.8.9/WithdrawalQueue.sol:292]: can be found on-chain also :)

File: contracts/0.8.9/WithdrawalQueue.sol [hunk 9]
```diff
@@ -203,92 +205,118 @@ contract WithdrawalQueue is AccessControlEnumerable, WithdrawalQueueBase {
...
-        whenResumed
-        returns (uint256 requestId)
-    {
-        _recipient = _checkWithdrawalRequestInput(IWstETH(WSTETH).getStETHByWstETH(_amountOfWstETH), _recipient);
-        return _requestWithdrawalWstETH(_amountOfWstETH, _recipient);
+    /// @notice Request the sequence of stETH withdrawals according to passed `withdrawalRequestInputs` data using EIP-2612 Permit
+    /// @param _withdrawalRequestInputs an array of `WithdrawalRequestInput` data. The standalone withdrawal request will
+    ///  be created for each item in the passed list. If `WithdrawalRequestInput.recipient` is set to `address(0)`,
+    ///  `msg.sender` will be used as recipient.
+    /// @return requestIds an array of the created withdrawal requests
```
Comment (TheDZhon) [inline contracts/0.8.9/WithdrawalQueue.sol:267]: permit param is undocumented

**Review by TheDZhon — APPROVED** [review]
LGTM mostly, left some additional suggestions

File: contracts/0.8.9/WithdrawalQueue.sol (hunk 1 above)
Comment (TheDZhon) [reply contracts/0.8.9/WithdrawalQueue.sol]: @DiRaiks fyi

File: contracts/0.8.9/WithdrawalQueue.sol [hunk 10]
```diff
@@ -203,92 +205,118 @@ contract WithdrawalQueue is AccessControlEnumerable, WithdrawalQueueBase {
...
+    function claimWithdrawals(ClaimWithdrawalInput[] calldata _claimWithdrawalInputs) external {
+        for (uint256 i = 0; i < _claimWithdrawalInputs.length; ++i) {
+            claimWithdrawal(_claimWithdrawalInputs[i].requestId, _claimWithdrawalInputs[i].hint);
+        }
+    }
+
+    /// @notice Returns the list of hints for the given request ids
+    /// @param _requestIds ids of the requests sorted in the descending order to get hints for
+    /// @return hintIds the hints for `claimWithdrawal` to find the discount for the passed request ids
+    function findClaimHints(uint256[] calldata _requestIds, uint256 _firstIndex, uint256 _lastIndex)
```
[the remark this answers is not in the document]
Comment (folkyatina) [reply contracts/0.8.9/WithdrawalQueue.sol:307]: They was not, I think. But thanks for the catch.

File: contracts/0.8.9/WithdrawalQueue.sol (hunk 3 above)
Comment (folkyatina) [reply contracts/0.8.9/WithdrawalQueue.sol:316]: Yep, it seems to be more intuitive. Agreed.

File: contracts/0.8.9/WithdrawalQueue.sol (hunk 9 above)
Comment (Psirex) [reply contracts/0.8.9/WithdrawalQueue.sol:268]: Added the doc comment for this param though maybe it is not very detailed.

File: contracts/0.8.9/WithdrawalQueue.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (Psirex) [reply contracts/0.8.9/WithdrawalQueue.sol:312]: Fixed. Added the doc comments for both params

**Pull request merged**